RV64 mscratch task ptr

furquan-rivos opened the pull request.

furquan-rivos committed:
sys/kern/rv64: Store current task pointer in mscratch

This change drops the global variable CURRENT_TASK_PTR and instead
stores the current task pointer in mscratch register. This drops
the usage of the variable in different functions and hence makes it
easier to split the files into separate submodules without having to
explicitly share this variable. This is in preparation to refactor
rv64 support into separate submodules in the future.

TEST=Verified that tests still pass for rv64.

Signed-off-by: Furquan Shaikh <furquan@rivosinc.com>

furquan-rivos committed:
sys/kern/rv64: Ensure mscratch is set before using `get_current_task`

This change adds an assert if `get_current_task()` is called without
mscratch being set.

adurbin-rivos commented on sys/kern/src/arch/riscv64.rs:
```diff
@@ -281,6 +273,7 @@ pub unsafe extern "C" fn _start_trap() {
         sd a1,  31*8(a0)    # store mepc for resume
         csrr a1, mscratch
         sd a1, 9*8(a0)      # store a0 itself
+        csrw mscratch, a0   # restore task ptr in mscratch
```
Not sure how much it all matters, but presumably you could have made line 274 csrrw a1, mscratch, a0.

furquan-rivos replied:
Good point. I can go ahead and push that as a follow-up change in the PR.

furquan-rivos replied:
> Is there a humility change needed for this? CURRENT_TASK_PTR can no longer be used. 

Yes, there is a change required in humility. Drew raised a [bug](http://go/ji/SW-600) to track that.

> Similarly, do we now need to indicate somehow in _start_trap() that current task cannot be relied upon in mscratch between lines 240 and 276?

Are you thinking of conditions like exception happening somewhere in between lines 240-276?

furquan-rivos replied:
Pushed follow-up commit to address review comments.

adurbin-rivos replied:
I was just thinking any sort of breakpoint or whatever when we are in this sequence and humility wants to interrogate the current running task.

furquan-rivos replied:
Ah I see. That is interesting. I will make a note of this on the bug as well so that we can think about this when we start adding support for humility.

furquan-rivos replied:
Done.

beezow reviewed (dismissed):
Looks good.  This will need Humility support.  Made this [ticket](https://rivosinc.atlassian.net/browse/SW-600) to track.

adurbin-rivos reviewed (commented):
Is there a humility change needed for this? CURRENT_TASK_PTR can no longer be used. Similarly, do we now need to indicate somehow in _start_trap() that current task cannot be relied upon in  mscratch between lines 240 and 276?

furquan-rivos committed:
sys/kern/rv64: mscratch current task ptr updates

This change performs following updates based on review feedback:
1. Use csrrw to read mscratch into a1 and a0 into mscratch instead of
   separate csrr and csrw.
2. Add comments to indicate that mscratch temporarily doesn't point to
   current task while stashing the current task status in
   _start_trap.

TEST=Verified that tests still pass for rv64.

furquan-rivos committed:
Merge branch 'rivos/main-rv64' into dev/furquan/rv64-mscratch-task-ptr
